enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,475 @@
+     annotation `k8s.ovn.org/hybrid-overlay-distributed-router-gateway-mac`,
+     which is required by hybrid overlay.
+
+     The script will also add the node annotation
+     `k8s.ovn.org/hybrid-overlay-node-subnet`, according to the HostSubnet CR of
+     this node.
+
+   - If `br-ex` does exist, it means the node has been updated by MCO. So the
+     `ovnkube-node` can be started.
+
+     The script will also add the node annotation `k8s.ovn.org/node-subnets`
+     according to the HostSubnet CR of this node.
+
+#### Migration
+
+When the condition 'Progressing' of the network Cluster Operator is switched
+from `True` to `False`, the admin can kick off the migration from CNO by:
+
+```bash
+$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"spec":{"networkType":"OVNKubernetes"}'
+```
+
+CNO will update the `status.networkType` field of the `network.config` CR. When
+MCO renders the MachineConfig, some host network configuration services will be
+rendered differently for different CNIs. For example, the
+`ovs-configuration.service` will create the br-ex bridge if the
+`status.networkType` is `OVNKubernetes`; it will not otherwise. So the change of
+the `status.networkType` field will trigger MCO to apply a new MachineConfig to
+each node:
+
+1. MCO will 
+   - render the MachineConfig for each node.
+   - try to cordon and drain the nodes based on
+     `MachineConfigPools.spec.maxUnavailable`
+2. MCO will drain and reboot the node.
+3. When the node is up, br-ex will be created. It means the node is ready for
+   OVN Kubernetes. So the local zone OVN Kubernetes components can be started.
+   The ovnkube-node container wrapper script will also add the node annotation
+  `k8s.ovn.org/node-subnets`, according to the HostSubnet CR of this node.
+4. On the local node, OVN Kubernetes will add all SDN nodes as hybrid overlay
+   nodes to the local zone and all OVN nodes as OVN-IC remote zones. On a remote
+   OVN node, OVN Kubernetes will remove a hybrid overlay node and add an OVN-IC
+   remote zone.
+5. MCO will uncordon the node. Pods will be recreated on the node using
+   ovn-kubernetes as the default CNI plugin.
+
+The above process will be repeated for each node until all the nodes have been
+applied to the new MachineConfig and converted to OVN Kubernetes.
+
+#### Migration Cleanup
+
+Once migration is complete, the admin can trigger the migration cleanup by:
+
+```bash
+$ oc patch Network.operator.openshift.io cluster --type='merge' --patch '{"spec":{"migration": null, "defaultNetwork":{ "openshiftSDNConfig": null}}'
+```
+
+CNO will:
+- delete the openshift-sdn DaemonSets and the related resources
+  (CustomResources, ConfigMaps, etc.).
+- redeploy ovn-kubernetes in "normal" mode (no migration mode script, hybrid
+  overlay disabled, etc.).
+- remove the migration-related labels and node annotations from the nodes.
+
+#### Operation Procedure
```

Comment:
General feedback on the manual touchpoints, we need to permit this to be done for any managed openshift (OSD/ROSA/ARO) whenever customer is ready.  We won't have SRE manually executing these steps.  It needs some controller to manage.  Is there a reason why this isn't handled as part of the migration automatically?  We need a way to express intent to start migration and an indicator when it's done.  In the middle if there are problems sufficient signal needs to be present so alerts can surface to SRE (or ops teams for self managed).  Each alert needs associated with clear steps on how to triage and fix known issues.  Of course, not everything will be known, we'll discover new things.  But some starting point will be needed else we go in blind.

Reply:
These steps can be automated and integrated into OCM. But that is out of the scope of this enhancement. As @Makdaam and I agreed, the SD team shall do the necessary UI and other changes in OCM. The cluster network operator and machine config operator will report errors in the corresponding ClusterOperator CRs. The detailed troubleshooting steps will be in the OCP document when the code is ready. The enhancement only contains the design of the feature.